atomic_reactor/utils/resource_manager.py

```diff
@@ -6,193 +6,290 @@
+        except TypeError as ex:  # Not an object
+            logger.warning("Json load failed. Not an object. %s", ex)
+            return None
+        except TypeError as ex:  # Unable to decode the bytes
+            logger.warning("Json load failed. Unable to decode the bytes. %s", ex)
+            return None
+        return json_stdout
+
+    def available_slots(self) -> list[int]:
+        """
+        Returns list of available slots
+
+        Slots should be randomized here, for smoother locking, because there will be
+        hunger games when every pipeline will try to lock slot with lowest number first
+        """
+        logger.debug("Retrieve list of available slots on host %s", self._hostname)
+
+        with self._ssh_session() as client:
+            cmd = self.get_cmd_to_read_from_json(RESOURCE_INFO_JSON)
+            resource_info = self._ssh_run_remote_cmd_to_get_json(client, cmd)
+
+            available_slots = []
+            for slot_id in range(resource_info.get('max_slot_count', 1)):
+                cmd = self.get_cmd_to_read_from_json(f'slot_{slot_id}.json')
+                slot = self._ssh_run_remote_cmd_to_get_json(client, cmd)
+
+                if slot is None:  # Json load failed - file on remote host might be corrupted
+                    logger.warning("%s - slot %s - json load failed", resource_info['hostname'], slot_id)
+                elif slot:  # Non-empty json file was loaded
+                    prid = slot.get('prid')
+                    logger.debug("%s - slot %s is occupied by %s", resource_info['hostname'], slot_id, prid)
+                    self._validate_resource_lock(slot)
+                elif not slot:  # Empty json file was loaded
+                    logger.debug("%s - slot %s is available", resource_info['hostname'], slot_id)
+                    available_slots.append(slot_id)
+                else:  # Undefined state - should not happen
+                    logger.warning("%s - slot %s invalid state!", resource_info['hostname'], slot_id)
+
+        return random.shuffle(available_slots)
+
+    def get_cmd_to_read_from_json(self, file_name: str):
+        """ Create a command to retrieve content of json file from remote host """
+        return ['cat', os.path.join(RELATIVE_PATH, file_name)]
+
+    def get_cmd_to_write_empty_slot(self, slot_id: int):
+        """ Create a command to unlock slot on remote host """
+        return ['echo', '{}', '>', os.path.join(RELATIVE_PATH, f'slot_{slot_id}.json')]
+
+    def get_cmd_to_write_slot_in_progress(self, file_name: str, payload: str):
+        """ Create a command to lock slot for pipelinerun on remote host """
+        filepath = os.path.join(RELATIVE_PATH, file_name)
+        _cmd = (f'if grep -qw {{}} {filepath} ; then echo {quote(payload)} > '
+                f'{filepath} ;else exit 1; fi') #  TODO: Test with remote host that this is actually working with Paramiko
+        return f'flock {filepath} -c {quote(_cmd)}'
```

Comment:
this locks the file ONLY for duration of the command ......

you have to use file descriptor

1st shell:
$ exec 99>"lock.file"
$ flock -n 99
$ echo $?
0

2nd shell
$ exec 99>"lock.file"
$ flock -n 99
$ echo $?
1

1st shell:
$ flock -u -n 99

2nd shell 
$ flock -n 99
$ echo $?
0

but all that works ONLY if that shell is still active

Reply:
also keep in mind that file actually isn't locked itself, any process can write to it

Reply:
so basically what it does now is that it is using flock feature just for writing to the file .... (was that really the intent ?)

Reply:
This part was not introduced by myself, I would like to re-test it to be sure it's working properly. But Anyway it's not complete, beacuse`critical section` will be required here for locking and unlocking as I mentioned in another comment, but it will be added in a separate PR.

Reply:
> so basically what it does now is that it is using flock feature just for writing to the file .... (was that really the intent ?)

Yes, I believe it was. `flock` is there only to prevent race conditions while locking and unlocking. The file does not have to stay flock-ed for the entire duration of the podman build

Reply:
well that seems weird anyway such usage of flock

as I've said earlier, anything can still write to that file anyway